tubular/dates.py

```diff
@@ -10,7 +10,94 @@
 from tubular.base import BaseTransformer
 
 
-class DateDiffLeapYearTransformer(BaseTransformer):
+class BaseDateTransformer(BaseTransformer):
+    """
+    Transformer with date data checks needed by all transformers taking dates as input data.
+    """
+
+    def check_columns_are_date_or_datetime(self, X):
+        "Raise a type error if a column to be operated on is not a datetime.datetime or datetime.date object"
+        for col in self.columns:
+            if (
+                not pd.api.types.is_datetime64_any_dtype(X[col])
+                and pd.api.types.infer_dtype(X[col]) != "date"
+            ):
+                msg = f"{self.classname()}: {col} should be datetime64 or date type but got {X[col].dtype}"
+                raise TypeError(msg)
+
+    def cast_columns_to_datetime(self, X):
+        """
+        Check whether columns are datetime64 or date type. If not, an error will be raised. If one is datetime.date
+        it will be cast to datetime64.
+        """
+
+        self.check_columns_are_date_or_datetime(X)
+
+        temp = X.copy()
```

Comment:
Its quite memory inefficient to copy the entire dataframe, instead is all we need to copy the columns that have been passed into the transformer? ie:

temp = X[self.columns].copy()